code/cabal-audit/src/Distribution/Audit.hs

```diff
@@ -0,0 +1,246 @@
+import Distribution.Client.ProjectOrchestration
+  ( CurrentCommand (OtherCommand)
+  , ProjectBaseContext (ProjectBaseContext, cabalDirLayout, distDirLayout, localPackages, projectConfig)
+  , commandLineFlagsToProjectConfig
+  , establishProjectBaseContext
+  )
+import Distribution.Client.ProjectPlanning (rebuildInstallPlan)
+import Distribution.Client.Setup (defaultGlobalFlags)
+import Distribution.Types.PackageName (PackageName, unPackageName)
+import Distribution.Verbosity qualified as Verbosity
+import Distribution.Version (Version, versionNumbers)
+import GHC.Generics (Generic)
+import Options.Applicative
+import Security.Advisories (Advisory (..), Keyword (..), ParseAdvisoryError, printHsecId)
+import Security.Advisories.Cabal (ElaboratedPackageInfoAdvised, ElaboratedPackageInfoWith (MkElaboratedPackageInfoWith, elaboratedPackageVersion, packageAdvisories), matchAdvisoriesForPlan)
+import Security.Advisories.Convert.OSV qualified as OSV
+import Security.Advisories.Filesystem (listAdvisories)
+import System.Exit (exitFailure)
+import System.IO.Temp (withSystemTempDirectory)
+import System.Process (callProcess)
+import Validation (validation)
+
+data AuditException
+  = -- | parsing the advisory database failed
+    ListAdvisoryValidationError {originalFilePath :: FilePath, parseError :: [ParseAdvisoryError]}
+  | -- | to rethrow exceptions thrown by cabal during plan elaboration
+    CabalException {reason :: String, cabalException :: SomeException}
+  deriving stock (Show, Generic)
+
+instance Exception AuditException where
+  displayException = \case
+    ListAdvisoryValidationError dir errs ->
+      mconcat
+        [ "Listing the advisories in directory "
+        , dir
+        , " failed with: \n"
+        , show errs
+        ]
+    CabalException ctx (SomeException ex) ->
+      "cabal failed while "
+        <> ctx
+        <> ":\n"
+        <> displayException ex
+
+-- | the type of output that is chosen for the command
+data Output
+  = -- | write humand readable to stdout
+    HumanReadable
+  | -- | write as Osv format to the specified file
+    Osv {osvPath :: FilePath}
+
+-- | configuration that is specific to the cabal audit command
+data AuditConfig = MkAuditConfig
+  { advisoriesPathOrURL :: Either FilePath String
+  -- ^ path or URL to the advisories
+  , verbosity :: Verbosity.Verbosity
+  -- ^ verbosity of cabal
+  , outputType :: Output
+  }
+
+-- | the main action to invoke
+auditMain :: IO ()
+auditMain =
+  do
+    (auditConfig, nixStyleFlags) <- customExecParser (prefs showHelpOnEmpty) do
+      info (helper <*> auditCommandParser) do
+        mconcat
+          [ fullDesc
+          , progDesc (formatWith [blue] "audit your cabal projects for vulnerabilities")
+          , header (formatWith [bold, blue] "Welcome to cabal audit")
+          ]
+
+    buildAdvisories auditConfig nixStyleFlags
+      >>= handleBuiltAdvisories (outputType auditConfig)
+    `catch` \(SomeException ex) -> do
+      putStrLn $
```

Comment:
This should go to stderr, not to stdout.

Reply:
Thank you, oversaw that.